File: platform/core/src/main/java/com/domatar/install/AssetPaths.java
/*
 * Copyright (c) 2024 Domatar
 */

package com.domatar.install;

/**
 * Browser URLs for any file served by an app's {@code AppAssetServlet}
 * (HTML, images, icons). Icons are paths under {@code icons/}, not a
 * separate URL language.
 *
 * <p>{@code url(origin, contextPath, appId, assetPath)} is the join.
 * {@code contextPath} is how browsers reach the WAR ({@code /domatar},
 * {@code ""}, or another prefix) — an install property, not a proxy brand.
 */
public final class AssetPaths
{
  public static final String WIRE_CONTEXT = "/domatar";
  public static final String DEFAULT_APP_ID = "domatar";
  public static final String LAUNCHER_ASSET = "icons/app.svg";
  public static final String DEFAULT_GLYPH_ASSET = "icons/cls/default/obj.svg";

  public static final String DEFAULT_RELATIVE =
      WIRE_CONTEXT + "/" + DEFAULT_APP_ID + "/" + DEFAULT_GLYPH_ASSET;

  private AssetPaths() {}

  /**
   * Browser URL for a file under {@code <appId>/assets/<assetPath>}.
   *
   * @param origin       {@code {scheme}://{host[:port]}} or null/empty
   *                     for a same-origin relative URL
   * @param contextPath  WAR mount as browsers see it ({@code /domatar}
   *                     or {@code ""}); null means {@link #WIRE_CONTEXT}
   * @param appId        app servlet mapping
   * @param assetPath    path inside AssetDirectory
   *                     ({@code icons/app.svg}, {@code img/a/b.png})
   */
  public static String url(final String origin, final String contextPath,
                           final String appId, final String assetPath)
  {
    if (appId == null || appId.isEmpty()
        || assetPath == null || assetPath.isEmpty())
      return url(origin, contextPath, DEFAULT_APP_ID, DEFAULT_GLYPH_ASSET);

    final String path = join(contextPath, appId, assetPath);

    if (origin == null || origin.isEmpty())
      return path;

    return trimTrailingSlash(origin) + path;
  }

  public static String launcher(final String origin, final String contextPath,
                                final String appId)
  {
    return url(origin, contextPath, appId, LAUNCHER_ASSET);
  }

  /** Wire-relative launcher ({@code /domatar/<appId>/icons/app.svg}). */
  public static String launcher(final String appId)
  {
    return launcher(null, WIRE_CONTEXT, appId);
  }

  public static String cls(final String origin, final String contextPath,
                           final String appId, final String clsId)
  {
    if (clsId == null || clsId.isEmpty())
      return url(origin, contextPath, DEFAULT_APP_ID, DEFAULT_GLYPH_ASSET);

    return url(origin, contextPath, appId, "icons/cls/" + clsId + ".svg");
  }

  public static String cls(final String appId, final String clsId)
  {
    return cls(null, WIRE_CONTEXT, appId, clsId);
  }

  public static String defaultGlyph(final String origin, final String contextPath)
  {
    return url(origin, contextPath, DEFAULT_APP_ID, DEFAULT_GLYPH_ASSET);
  }

  public static String defaultCls()
  {
    return DEFAULT_RELATIVE;
  }

  /**
   * Context browsers should use for this pair of hosts. Distinct public
   * host → empty context (front door remounts at {@code /}). Same host
   * or public unset → {@link #WIRE_CONTEXT}. Override with
   * {@code AssetContextPath} / {@code DOMATAR_ASSET_CONTEXT_PATH}.
   */
  public static String contextPath(final String publicDomain,
                                   final String wireDomain)
  {
    return usesFrontDoorPaths(publicDomain, wireDomain) ? "" : WIRE_CONTEXT;
  }

  public static String normalizeContext(final String contextPath)
  {
    if (contextPath == null)
      return WIRE_CONTEXT;

    String ctx = contextPath.trim();

    if (ctx.isEmpty() || "-".equals(ctx) || "none".equals(ctx)
        || "/".equals(ctx))
      return "";

    if (!ctx.startsWith("/"))
      ctx = "/" + ctx;

    while (ctx.length() > 1 && ctx.endsWith("/"))
      ctx = ctx.substring(0, ctx.length() - 1);

    return ctx;
  }

  public static String absolute(final String scheme, final String domain,
                                final String relativePath)
  {
    final String path = (relativePath == null || relativePath.isEmpty())
        ? DEFAULT_RELATIVE : relativePath;
    final String sch = (scheme == null || scheme.isEmpty()) ? "https" : scheme;
    final String dom = domain == null ? "" : domain;

    return sch + "://" + dom + path;
  }

  public static String stripContext(final String relativeWithContext)
  {
    if (relativeWithContext == null || relativeWithContext.isEmpty())
      return relativeWithContext;

    if (relativeWithContext.equals(WIRE_CONTEXT))
      return "/";

    if (relativeWithContext.startsWith(WIRE_CONTEXT + "/"))
      return relativeWithContext.substring(WIRE_CONTEXT.length());

    return relativeWithContext;
  }

  public static boolean usesFrontDoorPaths(final String publicDomain,
                                           final String wireDomain)
  {
    if (publicDomain == null || publicDomain.isEmpty())
      return false;

    if (wireDomain == null || wireDomain.isEmpty())
      return true;

    return !publicDomain.equals(wireDomain);
  }

  public static String browserAbsolute(final String scheme,
                                       final String publicDomain,
                                       final String wireDomain,
                                       final String relativeWithContext)
  {
    final String path = (relativeWithContext == null
        || relativeWithContext.isEmpty())
        ? DEFAULT_RELATIVE : relativeWithContext;

    if (usesFrontDoorPaths(publicDomain, wireDomain))
      return absolute(scheme, publicDomain, stripContext(path));

    final String host = (wireDomain != null && !wireDomain.isEmpty())
        ? wireDomain : publicDomain;

    return absolute(scheme, host, path);
  }

  public static String launcherAbsolute(final String scheme, final String domain,
                                        final String appId)
  {
    return absolute(scheme, domain, launcher(appId));
  }

  public static String clsAbsolute(final String scheme, final String domain,
                                   final String appId, final String clsId)
  {
    return absolute(scheme, domain, cls(appId, clsId));
  }

  public static String forInstall(final boolean remote, final String scheme,
                                  final String publicDomain,
                                  final String wireDomain, final String appId)
  {
    if (!remote)
      return launcher(appId);

    return browserAbsolute(scheme, publicDomain, wireDomain, launcher(appId));
  }

  public static String forInstall(final boolean remote, final String scheme,
                                  final String domain, final String appId)
  {
    return forInstall(remote, scheme, domain, domain, appId);
  }

  public static String assetOrigin(final String scheme, final String domain)
  {
    if (domain == null || domain.isEmpty())
      return null;

    final String sch = (scheme == null || scheme.isEmpty()) ? "https" : scheme;

    return sch + "://" + domain;
  }

  public static boolean isIconPath(final String pathInfo)
  {
    if (pathInfo == null)
      return false;

    return pathInfo.equals("/icons") || pathInfo.startsWith("/icons/");
  }

  /** Cache-Control/ETag candidates: {@code /icons/} and image files. */
  public static boolean isCachedAsset(final String pathInfo)
  {
    if (isIconPath(pathInfo))
      return true;

    if (pathInfo == null)
      return false;

    final String lower = pathInfo.toLowerCase();

    return lower.endsWith(".svg") || lower.endsWith(".png")
        || lower.endsWith(".jpg") || lower.endsWith(".jpeg")
        || lower.endsWith(".gif") || lower.endsWith(".webp")
        || lower.endsWith(".ico");
  }

  static String join(final String contextPath, final String appId,
                     final String assetPath)
  {
    final String ctx = (contextPath == null)
        ? WIRE_CONTEXT : normalizeContext(contextPath);
    final String asset = trimLeadingSlash(assetPath.replace('\\', '/'));

    return ctx + "/" + appId + "/" + asset;
  }

  private static String trimLeadingSlash(final String path)
  {
    String p = path;

    while (p.startsWith("/"))
      p = p.substring(1);

    return p;
  }

  private static String trimTrailingSlash(final String origin)
  {
    if (origin.endsWith("/"))
      return origin.substring(0, origin.length() - 1);

    return origin;
  }
}

File: platform/war/src/main/webapp/asset-paths.js
/*
 * Copyright (c) 2024 Domatar
 *
 * Browser URLs for any file served by an app's AppAssetServlet
 * (HTML, images, icons). Mirrors com.domatar.install.AssetPaths.
 *
 * WIRE_CONTEXT is built with concatenation so a reverse-proxy sub_filter
 * on "/domatar/" cannot rewrite this file's source.
 */
(function (global)
{
    var WIRE_CONTEXT = "/" + "domatar";
    var DEFAULT_APP_ID = "domatar";
    var LAUNCHER_ASSET = "icons/app.svg";
    var DEFAULT_GLYPH_ASSET = "icons/cls/default/obj.svg";
    var SCRIPT_NAME = "asset-paths.js";

    function normalizeContext(contextPath)
    {
        if (contextPath == null)
            return WIRE_CONTEXT;

        var ctx = String(contextPath).replace(/^\s+|\s+$/g, "");

        if (ctx === "" || ctx === "-" || ctx === "none" || ctx === "/")
            return "";

        if (ctx.charAt(0) !== "/")
            ctx = "/" + ctx;

        while (ctx.length > 1 && ctx.charAt(ctx.length - 1) === "/")
            ctx = ctx.substring(0, ctx.length - 1);

        return ctx;
    }

    function trimLeadingSlash(path)
    {
        var p = String(path || "").replace(/\\/g, "/");

        while (p.charAt(0) === "/")
            p = p.substring(1);

        return p;
    }

    function join(contextPath, appId, assetPath)
    {
        var ctx = (contextPath == null) ? WIRE_CONTEXT : normalizeContext(contextPath);
        var asset = trimLeadingSlash(assetPath);

        return ctx + "/" + appId + "/" + asset;
    }

    function url(origin, contextPath, appId, assetPath)
    {
        if (!appId || !assetPath)
            return url(origin, contextPath, DEFAULT_APP_ID, DEFAULT_GLYPH_ASSET);

        var path = join(contextPath, appId, assetPath);

        if (!origin)
            return path;

        var base = String(origin);
        if (base.charAt(base.length - 1) === "/")
            base = base.substring(0, base.length - 1);

        return base + path;
    }

    function contextFromScript()
    {
        var scripts = document.getElementsByTagName("script");
        var i, src, idx, u, prefix;

        for (i = 0; i < scripts.length; i++)
        {
            src = scripts[i].src || "";
            idx = src.indexOf("/" + SCRIPT_NAME);
            if (idx < 0 && src.indexOf(SCRIPT_NAME) >= 0)
                idx = src.lastIndexOf("/");
            if (idx < 0)
                continue;
            try
            {
                u = new URL(src, global.location.href);
                prefix = u.pathname.substring(0, u.pathname.lastIndexOf("/"));
                if (prefix === WIRE_CONTEXT || prefix.indexOf(WIRE_CONTEXT + "/") === 0)
                    return WIRE_CONTEXT;
                return normalizeContext(prefix === "/" ? "" : prefix);
            }
            catch (e) {}
        }

        return "";
    }

    function sameHost(origin)
    {
        if (!origin)
            return true;
        try
        {
            return (new URL(origin, global.location.href)).host === global.location.host;
        }
        catch (e)
        {
            return false;
        }
    }

    /**
     * @param {string} [origin]       scheme://host or empty for this page
     * @param {string|null} [contextPath]  /domatar, "", or null to infer
     *        from how this script was loaded
     */
    function AssetPaths(origin, contextPath)
    {
        this.origin = origin || "";
        this.contextPath = (contextPath === undefined || contextPath === null)
            ? contextFromScript()
            : normalizeContext(contextPath);
    }

    AssetPaths.WIRE_CONTEXT = WIRE_CONTEXT;
    AssetPaths.DEFAULT_APP_ID = DEFAULT_APP_ID;
    AssetPaths.LAUNCHER_ASSET = LAUNCHER_ASSET;
    AssetPaths.DEFAULT_GLYPH_ASSET = DEFAULT_GLYPH_ASSET;

    AssetPaths.normalizeContext = normalizeContext;
    AssetPaths.url = url;
    AssetPaths.contextFromScript = contextFromScript;

    AssetPaths.fromSiteRole = function (role)
    {
        if (!role || role.AssetContextPath === undefined || role.AssetContextPath === null)
            return new AssetPaths("", null);

        return new AssetPaths("", role.AssetContextPath);
    };

    AssetPaths.fromHosts = function (domain, publicDomain)
    {
        var host = publicDomain || domain || "";

        if (!host)
            return new AssetPaths("", null);

        var origin = host.indexOf("://") >= 0
            ? host
            : (global.location.protocol + "//" + host);

        if (sameHost(origin))
            return new AssetPaths("", null);

        var ctx = (publicDomain && publicDomain !== domain) || (publicDomain && !domain)
            ? ""
            : WIRE_CONTEXT;

        return new AssetPaths(origin, ctx);
    };

    AssetPaths.prototype.url = function (appId, assetPath)
    {
        var origin = this.origin;
        var ctx = this.contextPath;
        var path = (global.location && global.location.pathname) || "";
        var onWar = path === WIRE_CONTEXT
            || path.indexOf(WIRE_CONTEXT + "/") === 0;

        if (onWar)
        {
            var originIsLocal = false;
            try
            {
                if (origin)
                {
                    var h = (new URL(origin, global.location.href)).hostname;
                    originIsLocal = h === "localhost" || h === "127.0.0.1";
                }
            }
            catch (e) { originIsLocal = false; }

            if (origin && !sameHost(origin) && originIsLocal)
                ctx = WIRE_CONTEXT;
            else
            {
                origin = "";
                ctx = WIRE_CONTEXT;
            }
        }
        else if (sameHost(origin))
            origin = "";

        return url(origin, ctx, appId, assetPath);
    };

    AssetPaths.prototype.launcher = function (appId)
    {
        return this.url(appId, LAUNCHER_ASSET);
    };

    AssetPaths.prototype.cls = function (appId, clsId)
    {
        if (!clsId)
            return this.defaultGlyph();

        return this.url(appId, "icons/cls/" + clsId + ".svg");
    };

    AssetPaths.prototype.defaultGlyph = function ()
    {
        return this.url(DEFAULT_APP_ID, DEFAULT_GLYPH_ASSET);
    };

    AssetPaths.prototype.isDefaultGlyph = function (src)
    {
        if (!src)
            return false;

        var def = this.defaultGlyph();

        return src.indexOf(def) !== -1
            || src.indexOf("/" + DEFAULT_APP_ID + "/" + DEFAULT_GLYPH_ASSET) !== -1
            || src.indexOf("/icons/cls/default/obj.svg") !== -1;
    };

    global.AssetPaths = AssetPaths;
})(window);
